### ISSUE TITLE
ideas re: SDXL support for ClownRegionalConditioning3?

### ISSUE THREAD

**Issue by jhlkbrks** [ClownsharkBatwing/RES4LYF#94]
Big fan, incredible work you have done here! Do you have any pointers or ideas for getting your ClownRegionalConditioning3 node working in an SDXL workflow? It passes everything but doesn't actually seem to affect it, & I can't figure out why. Thanks!
conditioning.py:
class ClownRegionalConditioning3:
@classmethod
def INPUT_TYPES(cls):
    return {
        "required": { 
            "weight":                  ("FLOAT",                                     {"default": 1.0,  "min": -10000.0, "max": 10000.0, "step": 0.01}),
            "region_bleed":            ("FLOAT",                                     {"default": 0.0,  "min": -10000.0, "max": 10000.0, "step": 0.01}),
            "region_bleed_start_step": ("INT",                                       {"default": 0,   "min":  0,        "max": 10000}),
            "weight_scheduler":        (["constant"] + get_res4lyf_scheduler_list(), {"default": "constant"},),
            "start_step":              ("INT",                                       {"default": 0,    "min":  0,        "max": 10000}),
            "end_step":                ("INT",                                       {"default": 100,  "min": -1,        "max": 10000}),
            "mask_type":               (["gradient", "gradient_masked", "gradient_unmasked", "boolean", "boolean_masked", "boolean_unmasked"],                     {"default": "boolean"}),
            "narcissism_area":         (["A", "B", "AB", "unmasked", "off"],         {"default": "off"}),
            "narcissism_start_step":   ("INT",                                 {"default": 0,   "min": -1,        "max": 10000}),
            "narcissism_end_step":     ("INT",                                 {"default": 5,   "min": -1,        "max": 10000}),
            "invert_mask":             ("BOOLEAN",                                   {"default": False}),
        }, 
        "optional": {
            "positive_A":              ("CONDITIONING", ),
            "positive_B":              ("CONDITIONING", ),
            "positive_unmasked":       ("CONDITIONING", ),
            "mask_A":                  ("MASK", ),
            "mask_B":                  ("MASK", ),
            "weights":                 ("SIGMAS", ),
            "region_bleeds":           ("SIGMAS", ),
        }
    }
RETURN_TYPES = ("CONDITIONING",)
RETURN_NAMES = ("positive",)
FUNCTION     = "main"
CATEGORY     = "RES4LYF/conditioning"
def create_callback(self, **kwargs):
    def callback(model):
        kwargs["model"] = model  
        pos_cond, = self.prepare_regional_cond(**kwargs)
        return pos_cond
    return callback
def main(self,
        weight                   : float  = 1.0,
        start_percent            : float  = 0.0,
        end_percent              : float  = 1.0,
        weight_scheduler                  = None,
        start_step               : int    = 0,
        end_step                 : int    = -1,
        positive_A                   = None,
        positive_B                        = None,
        positive_unmasked                 = None,
        weights                  : Tensor = None,
        region_bleeds            : Tensor = None,
        region_bleed             : float  = 0.0,
        region_bleed_start_step  : int    = 0,
        mask_type                : str    = "boolean",
        mask_A                              = None,
        mask_B                              = None,
        narcissism_area       : str    = "AB",
        narcissism_start_step : int    = 0,
        narcissism_end_step   : int    = 5,
        invert_mask              : bool   = False
        ) -> Tuple[Tensor]:
    
    if end_step == -1:
        end_step = MAX_STEPS
    
    callback = self.create_callback(weight                   = weight,
                                    start_percent            = start_percent,
                                    end_percent              = end_percent,
                                    weight_scheduler         = weight_scheduler,
                                    start_step               = start_step,
                                    end_step                 = end_step,
                                    weights                  = weights,
                                    region_bleeds            = region_bleeds,
                                    region_bleed             = region_bleed,
                                    region_bleed_start_step  = region_bleed_start_step,
                                    mask_type                = mask_type,
                                    mask_A                   = mask_A,
                                    mask_B                   = mask_B,
                                    invert_mask              = invert_mask,
                                    positive_A               = positive_A,
                                    positive_B               = positive_B,
                                    positive_unmasked        = positive_unmasked,
                                    narcissism_area       = narcissism_area,
                                    narcissism_start_step = narcissism_start_step,
                                    narcissism_end_step   = narcissism_end_step,
                                    )
    pooled_len = 768
    if positive_A is not None:
        pooled     = positive_A[0][1].get('pooled_output')
        pooled_len = pooled.shape[-1] if pooled is not None else pooled_len
        positive = [[
            torch.zeros_like(positive_A[0][0]),
            {"pooled_output": torch.zeros( (1,pooled_len), dtype=positive_A[0][0].dtype, device=positive_A[0][0].device  )},
            #{}
            #{"pooled_output": torch.zeros_like(positive_masked[0][1]['pooled_output'])}
        ]]
        
    if positive_B is not None:
        pooled     = positive_B[0][1].get('pooled_output')
        pooled_len = pooled.shape[-1] if pooled is not None else pooled_len
        positive = [[
            torch.zeros_like(positive_B[0][0]),
            {"pooled_output": torch.zeros( (1,pooled_len), dtype=positive_B[0][0].dtype, device=positive_B[0][0].device  )},
            #{}
            #{"pooled_output": torch.zeros_like(positive_masked[0][1]['pooled_output'])}
        ]]
        
    elif positive_unmasked is not None:
        pooled     = positive_unmasked[0][1].get('pooled_output')
        pooled_len = pooled.shape[-1] if pooled is not None else pooled_len
        positive = [[
            torch.zeros_like(positive_unmasked[0][0]),
            {"pooled_output": torch.zeros( (1,pooled_len), dtype=positive_unmasked[0][0].dtype, device=positive_unmasked[0][0].device  )},
            #{"pooled_output": torch.zeros_like(positive_unmasked[0][1]['pooled_output'])}
        ]]
    """positive = [[
        torch.zeros((1, 256, 4096)),
        {'pooled_output': torch.zeros((1, 768))}
        ]]"""
    
    positive[0][1]['callback_regional'] = callback
    
    return (positive,)
def prepare_regional_cond(self,
                            model,
                            weight                   : float  = 1.0,
                            start_percent            : float  = 0.0,
                            end_percent              : float  = 1.0,
                            weight_scheduler                  = None,
                            start_step               : int    =  0,
                            end_step                 : int    = -1,
                            positive_A                        = None,
                            positive_B                        = None,
                            positive_unmasked                 = None,
                            weights                  : Tensor = None,
                            region_bleeds            : Tensor = None,
                            region_bleed             : float  = 0.0,
                            region_bleed_start_step  : int    = 0,
                            mask_type                : str    = "boolean",
                            mask_A                            = None,
                            mask_B                            = None,
                            invert_mask              : bool   = False,
                            narcissism_area       : str    = "AB",
                            narcissism_start_step : int    = 0,
                            narcissism_end_step   : int    = 5,
                            ) -> Tuple[Tensor]:
    default_dtype  = torch.float64
    default_device = torch.device("cuda") 
    
    if end_step == -1:
        end_step = MAX_STEPS
    
    if weights is None and weight_scheduler != "constant":
        total_steps = end_step - start_step
        weights     = get_sigmas(model, weight_scheduler, total_steps, 1.0).to(dtype=default_dtype, device=default_device) #/ model.inner_model.inner_model.model_sampling.sigma_max  #scaling doesn't matter as this is a flux-only node
        prepend     = torch.zeros(start_step,                                  dtype=default_dtype, device=default_device)
        weights     = torch.cat((prepend, weights), dim=0)
    
    if invert_mask and mask_A is not None:
        mask_A = 1-mask_A
        
    if invert_mask and mask_B is not None:
        mask_B = 1-mask_B
    
    mask_AB_inv = torch.ones_like(mask_A) - mask_A - mask_B
    #weight, weights = mask_weight, mask_weights
    floor, floors = region_bleed, region_bleeds
    
    weights = initialize_or_scale(weights, weight, end_step).to(default_dtype)
    weights = F.pad(weights, (0, MAX_STEPS), value=0.0)
    
    prepend    = torch.full((region_bleed_start_step,),  0.0, dtype=default_dtype, device=default_device)
    floors  = initialize_or_scale(floors,  floor,  end_step).to(default_dtype).to(default_device)
    floors  = F.pad(floors,  (0, MAX_STEPS), value=0.0)
    floors  = torch.cat((prepend, floors), dim=0)
    if (positive_A is None) and (positive_B is None) and (positive_unmasked is None):
        positive = None
    elif mask_A is not None:
        if   isinstance(model.model.model_config, comfy.supported_models.SD3):
            text_len_base = 154
            pooled_len    = 2048
            text_channels = 4096
        elif isinstance(model.model.model_config, comfy.supported_models.Flux) \
            or isinstance(model.model.model_config, comfy.supported_models.FluxSchnell):
            text_len_base = 256
            pooled_len    = 768
            text_channels = 4096
        elif isinstance(model.model.model_config, comfy.supported_models.AuraFlow):
            text_len_base = 256
            pooled_len    = 768
            text_channels = 2048
        elif isinstance(model.model.model_config, comfy.supported_models.Stable_Cascade_C):
            text_len_base = 77
            pooled_len    = 1280
            text_channels = 1280
        elif isinstance(model.model.model_config, comfy.supported_models.SDXL):
            text_len_base = 231
            pooled_len    = 1280
            text_channels = 2048
        
        
        #elif isinstance(model.model.model_config, comfy.supported_models.Cascade_StageC):
        if positive_A is None:    
            if positive_A is None:
                positive_A = [[
                    torch.zeros((1, text_len_base, text_channels)),
                    {'pooled_output': torch.zeros((1, pooled_len))}
                    ]]
        if positive_B is None:    
            if positive_B is None:
                positive_B = [[
                    torch.zeros((1, text_len_base, text_channels)),
                    {'pooled_output': torch.zeros((1, pooled_len))}
                    ]]
        if positive_unmasked is None:    
            if positive_unmasked is None:
                positive_unmasked = [[
                    torch.zeros((1, text_len_base, text_channels)),
                    {'pooled_output': torch.zeros((1, pooled_len))}
                    ]]
        positive_A_tokens        = positive_A[0][0].shape[1]
        positive_B_tokens        = positive_B[0][0].shape[1]
        positive_unmasked_tokens = positive_unmasked[0][0].shape[1]
        
        values = sorted([positive_A_tokens, positive_B_tokens, positive_unmasked_tokens])
        positive_min_tokens  = values[0]
        positive_mid_tokens  = values[1]
        positive_max_tokens  = values[2]
        
        positive = copy.deepcopy(positive_A)
        positive[0][0] = (positive_A[0][0][:,:positive_min_tokens,:] + positive_B[0][0][:,:positive_min_tokens,:] + positive_unmasked[0][0][:,:positive_min_tokens,:]) / 3
        
        from .attention_masks import FullAttentionMask, SplitAttentionMask, RegionalContext
        if isinstance(model.model.model_config, comfy.supported_models.WAN21_T2V) or isinstance(model.model.model_config, comfy.supported_models.WAN21_I2V) or isinstance(model.model.model_config, comfy.supported_models.SDXL):
            AttnMask = SplitAttentionMask(mask_type)
        else:
            AttnMask = FullAttentionMask(mask_type)
        AttnMask.add_region(positive_A       [0][0], mask_A)
        AttnMask.add_region(positive_B       [0][0], mask_B)
        AttnMask.add_region(positive_unmasked[0][0], mask_AB_inv)
        #AttnMask.generate()
        
        RegContext = RegionalContext()
        RegContext.add_region(positive_A  [0][0])
        RegContext.add_region(positive_B[0][0])
        RegContext.add_region(positive_unmasked[0][0])
        
        positive[0][1]['AttnMask']   = AttnMask
        positive[0][1]['RegContext'] = RegContext
        
        
        if   positive_A_tokens        != positive_min_tokens and positive_A_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_A       [0][0][:,positive_min_tokens:,:]), dim=1)
            
        elif positive_B_tokens        != positive_min_tokens and positive_B_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_B       [0][0][:,positive_min_tokens:,:]), dim=1)
            
        elif positive_unmasked_tokens != positive_min_tokens and positive_unmasked_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_unmasked[0][0][:,positive_min_tokens:,:]), dim=1)
            
            
            
        if   positive_A_tokens        == positive_mid_tokens and positive_mid_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_A       [0][0][:,positive_mid_tokens:,:]), dim=1)
            
        elif positive_B_tokens        == positive_mid_tokens and positive_mid_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_B       [0][0][:,positive_mid_tokens:,:]), dim=1)
            
        elif positive_unmasked_tokens == positive_mid_tokens and positive_mid_tokens != positive_max_tokens:
            positive[0][0] = torch.cat((positive[0][0], positive_unmasked[0][0][:,positive_mid_tokens:,:]), dim=1)
        
        
        
        if 'pooled_output' in positive[0][1] and positive_A[0][1]['pooled_output'] is not None:
            positive_A_pooled_tokens = positive_A[0][1]['pooled_output'].shape[1]
            positive_B_pooled_tokens = positive_B[0][1]['pooled_output'].shape[1]
            positive_unmasked_pooled_tokens = positive_unmasked[0][1]['pooled_output'].shape[1]
            
            values = sorted([positive_A_pooled_tokens, positive_B_pooled_tokens, positive_unmasked_pooled_tokens])
            positive_min_pooled_tokens  = values[0]
            positive_mid_pooled_tokens  = values[1]
            positive_max_pooled_tokens  = values[2]
            
            positive[0][1]['pooled_output'] = (positive_A[0][1]['pooled_output'][:,:positive_min_pooled_tokens] + positive_B[0][1]['pooled_output'][:,:positive_min_pooled_tokens] + positive_unmasked[0][1]['pooled_output'][:,:positive_min_pooled_tokens]) / 3
            
            if   positive_A_pooled_tokens        != positive_min_pooled_tokens and positive_A_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_A       [0][1]['pooled_output'][:,positive_min_pooled_tokens:,:]), dim=1)
                
            elif positive_B_pooled_tokens        != positive_min_pooled_tokens and positive_B_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_B       [0][1]['pooled_output'][:,positive_min_pooled_tokens:,:]), dim=1)
                
            elif positive_unmasked_pooled_tokens != positive_min_pooled_tokens and positive_unmasked_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_unmasked[0][1]['pooled_output'][:,positive_min_pooled_tokens:,:]), dim=1)
                
                
                
            if   positive_A_pooled_tokens        == positive_mid_pooled_tokens and positive_mid_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_A       [0][1]['pooled_output'][:,positive_mid_pooled_tokens:,:]), dim=1)
                
            elif positive_B_pooled_tokens        == positive_mid_pooled_tokens and positive_mid_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_B       [0][1]['pooled_output'][:,positive_mid_pooled_tokens:,:]), dim=1)
                
            elif positive_unmasked_pooled_tokens == positive_mid_pooled_tokens and positive_mid_pooled_tokens != positive_max_pooled_tokens:
                positive[0][1]['pooled_output'] = torch.cat((positive[0][1]['pooled_output'], positive_unmasked[0][1]['pooled_output'][:,positive_mid_pooled_tokens:,:]), dim=1)
            
    else:
        positive = positive_A
    
    if   mask_A is not None and narcissism_area == "A":
        narc_idx = 0
        positive[0][1]['regional_conditioning_mask_orig'] = 1-mask_A.clone()
        
    elif mask_B is not None and narcissism_area == "B":
        narc_idx = 1
        positive[0][1]['regional_conditioning_mask_orig'] = 1-mask_B.clone()
        
    elif mask_A is not None and mask_B is not None and narcissism_area == "AB":
        narc_idx = 2
        positive[0][1]['regional_conditioning_mask_orig'] = torch.clamp(1 - mask_A.clone() - mask_B.clone(), min=0.0)
        
    elif mask_A is not None and mask_B is not None and narcissism_area == "unmasked":
        narc_idx = 25436747  # need to deal with this properly!
        positive[0][1]['regional_conditioning_mask_orig'] = 1-torch.clamp(1 - mask_A.clone() - mask_B.clone(), min=0.0)
        
    else:
        narc_idx = -1
        positive[0][1]['regional_conditioning_mask_orig'] = None
    
    positive[0][1]['RegParam'] = RegionalParameters(weights, floors, narc_idx, narcissism_start_step, narcissism_end_step)
    
    return (positive,)

**Comment by ClownsharkBatwing** [comment]
I't something I'll probably do _eventually_ once I get caught up on a bunch of other things on my to-do list. It will require reworking the SDXL mode code to do, but there's no reason it can't be done.
